src/order/compactly_generated.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import order.well_founded
+import order.order_iso_nat
+import data.set.finite
+import tactic.tfae
+
+/-!
+# Compactness properties for complete lattices
+
+For complete lattices, there are numerous equivalent ways to express the fact that the relation `>`
+is well-founded. In this file we define three especially-useful characterisations and provide
+proofs that they are indeed equivalent to well-foundedness.
+
+## Main definitions
+ * `is_sup_closed_compact`
+ * `is_Sup_finite_compact`
+ * `is_compact_element`
+ * `is_compactly_generated`
+
+## Main results
+The main result is that the following four conditions are equivalent for a complete lattice:
+ * `well_founded (>)`
+ * `is_sup_closed_compact`
+ * `is_Sup_finite_compact`
+ * `∀ k, is_compact_element k`
+
+This is demonstrated by means of the following four lemmas:
+ * `well_founded.is_Sup_finite_compact`
+ * `is_Sup_finite_compact.is_sup_closed_compact`
+ * `is_sup_closed_compact.well_founded`
+ * `is_Sup_finite_compact_iff_all_elements_compact`
+
+ We also show well-founded lattices are compactly generated.
+
+## Tags
+
+complete lattice, well-founded, compact
+-/
+
+namespace complete_lattice
+
+variables (α : Type*) [complete_lattice α]
+
+/-- A compactness property for a complete lattice is that any `sup`-closed non-empty subset
+contains its `Sup`. -/
+def is_sup_closed_compact : Prop :=
+  ∀ (s : set α) (h : s.nonempty), (∀ a b, a ∈ s → b ∈ s → a ⊔ b ∈ s) → (Sup s) ∈ s
+
+/-- A compactness property for a complete lattice is that any subset has a finite subset with the
+same `Sup`. -/
+def is_Sup_finite_compact : Prop :=
+∀ (s : set α), ∃ (t : finset α), ↑t ⊆ s ∧ Sup s = t.sup id
+
+/-- An element k of a complete lattice is said to be compact if any set that Sups
+above k has a finite subset that Sups above k -/
```

Comment:
I moved some text from the docstring of `is_compact_element_iff_le_of_directed_Sup_le` here:
```suggestion
/-- An element `k` of a complete lattice is said to be compact if any set with `Sup`
above `k` has a finite subset with `Sup` above `k`.  Such an element is also called
"finite" or "S-compact". -/
```